packages/excubiae/contracts/Excubia.sol

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0 <0.9.0;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IExcubia} from "./IExcubia.sol";
+
+/// @title Excubia.
+/// @notice Abstract base contract which can be extended to implement a specific excubia.
+/// @dev Inherit from this contract and implement the `_check` method to define the custom gatekeeping logic.
+abstract contract Excubia is IExcubia, Ownable(msg.sender) {
+    /// @notice The excubia-protected contract address.
+    /// @dev The gate can be any contract address that requires a prior `_check`.
+    /// For example, the gate is a semaphore group that requires the passerby
+    /// to meet certain criteria before joining.
+    address public gate;
+
+    /// @notice Tracks whether an address has successfully passed the gate check.
+    /// @dev Maps an address to a boolean indicating pass status.
+    mapping(address => bool) public checked;
```

Comment:
so here's the tricky part, given this is supposed to be the abstract contract everyone inherits from, I think it's best that either this mapping is as generalized as possible, so instead of `address` you store the encoded data like `bytes` (which not sure if it always works), or you omit this mapping from this contract and make it clear that each gate contract should implement this mapping to prevent double signups. Just thinking along the lines of: a user comes to the gate with something they own, and that something that owns is what grants them access, like in a door you grant access to anyone that has a key. If you then keep an image of their face to prevent entering once, they could give the key to someone else to enter, but if you prevent the same key from entering, even if they give the key to someone else they won't be able to enter. This assumes the "key" can be transferred which might not always be the case in something like EAS, Zupass, etc. but just trying to make this as general as possible if that makes sense

Reply:
TLDR I think in general it might be easier if the gate specifies the data type of this mapping

Reply:
got it - check my proposal below

```solidity
mapping(bytes32 => bool) public checked;

function _pass(bytes memory data, address passerby) internal virtual onlyGate {
    // proposal
    bytes32 key = keccak256(abi.encode(passerby, data)); // creating a unique key (?)
    
    if (checked[key]) revert AccessAlreadyGranted();
    if (!_check(data, passerby)) revert AccessDenied();

    checked[key] = true;
    
    emit GatePassed(passerby, gate);
}
```

Reply:
got it - check my proposal below

```solidity
mapping(bytes32 => bool) public checked;

function _pass(bytes memory data, address passerby) internal virtual onlyGate {
    // proposal
    bytes32 key = keccak256(abi.encode(passerby, data)); // creating a unique key (?)
    
    if (checked[key]) revert AccessAlreadyGranted();
    if (!_check(data, passerby)) revert AccessDenied();

    checked[key] = true;
    
    emit GatePassed(passerby, gate);
}
```

Reply:
or we can have a `makeKey()` abstract method which returns the key in `bytes32`?

Reply:
OK, all clear. To stop someone else (a passerby) from using the same data, regardless of protocol, to try to pass the gatekeeper check twice but with a different address, we need to save two mappings: one for the passerby and one for the data.

In short, we have two options:
1. Do nothing and let the abstract contract implementer decide what information to save in the pass + check logic to avoid sybil.
2. Make sure everyone saves the address and data in mappings.

As we discussed in the chat, let's stick with the first option for now until we come up with a registry or logic that can be merged into a single solution.

Reply:
btw, reflected these decisions in the usage section of the README, specifying that the developer needs to provide their own logic for gatekeeping + preventing unwanted access.